src/goto-instrument/code_contracts.cpp

```diff
@@ -257,14 +229,74 @@ const symbolt &code_contractst::new_tmp_symbol(
     symbol_table);
 }
 
+bool code_contractst::enforce_contract(const std::string &fun_to_enforce)
+{
+  // Rename old function
+  std::stringstream ss;
+  ss << CPROVER_PREFIX << "contracts_original_" << fun_to_enforce;
+  const irep_idt mangled(ss.str());
+  const irep_idt original(fun_to_enforce);
+
+  auto old_fun = goto_functions.function_map.find(original);
+  if(old_fun == goto_functions.function_map.end())
+  {
+    log.error() << "Could not find function '" << fun_to_enforce
+                << "' in goto-program; not enforcing contracts."
+                << messaget::eom;
+    return true;
+  }
+  std::swap(goto_functions.function_map[mangled], old_fun->second);
+  goto_functions.function_map.erase(old_fun);
+
+  // Place a new symbol with the mangled name into the symbol table
+  source_locationt sl;
+  sl.set_file("instrumented for code contracts");
+  sl.set_line("0");
+  symbolt mangled_sym;
+  const symbolt *original_sym = symbol_table.lookup(original);
+  mangled_sym = *original_sym;
+  mangled_sym.name = mangled;
+  mangled_sym.base_name = mangled;
+  mangled_sym.location = sl;
+  const auto mangled_found = symbol_table.insert(std::move(mangled_sym));
+  INVARIANT(
+    mangled_found.second,
+    "There should be no existing function called " + ss.str() +
+      " in the symbol table because that name is a mangled name");
+
+  // Insert wrapper function into goto_functions
+  auto nexist_old_fun = goto_functions.function_map.find(original);
+  INVARIANT(
+    nexist_old_fun == goto_functions.function_map.end(),
+    "There should be no function called " + fun_to_enforce +
+      " in the function map because that function should have had its"
+      " name mangled");
+
+  auto mangled_fun = goto_functions.function_map.find(mangled);
+  INVARIANT(
```

Comment:
Many thanks for all the INVARIANT's - makes this a lot more readable/documented.